be/src/runtime/batch_write/txn_state_cache.cpp

```diff
@@ -0,0 +1,495 @@
+                              << ", tbl: " << task.tbl;
+        }
+    }
+}
+
+void TxnStatePoller::_execute_poll(const TxnStatePollTask& task) {
+    int64_t start_ts = MonotonicMicros();
+    TGetLoadTxnStatusRequest request;
+    request.__set_db(task.db);
+    request.__set_tbl(task.tbl);
+    request.__set_txnId(task.txn_id);
+    set_request_auth(&request, task.auth);
+    TGetLoadTxnStatusResult response;
+    Status status;
+#ifndef BE_TEST
+    TNetworkAddress master_addr = get_master_address();
+    status = ThriftRpcHelper::rpc<FrontendServiceClient>(
+            master_addr.hostname, master_addr.port,
+            [&request, &response](FrontendServiceConnection& client) { client->getLoadTxnStatus(response, request); });
+#else
+    TEST_SYNC_POINT_CALLBACK("TxnStatePoller::_execute_poll::request", &request);
+    TEST_SYNC_POINT_CALLBACK("TxnStatePoller::_execute_poll::status", &status);
+    TEST_SYNC_POINT_CALLBACK("TxnStatePoller::_execute_poll::response", &response);
+#endif
+    TRACE_BATCH_WRITE << "execute poll task, txn_id: " << task.txn_id << ", db: " << task.db << ", tbl: " << task.tbl
+                      << ", cost: " << (MonotonicMicros() - start_ts) << " us, rpc status: " << status
+                      << ", response: " << response;
+    if (status.ok()) {
+        _txn_state_cache->_notify_poll_result(task, TxnState{response.status, response.reason});
+    } else {
+        _txn_state_cache->_notify_poll_result(
+                task, Status::InternalError("poll txn state failed, error: " + status.to_string()));
+    }
+}
+
+bool TxnStatePoller::TEST_is_txn_pending(int64_t txn_id) {
+    std::unique_lock<bthread::Mutex> lock(_mutex);
+    return _pending_txn_ids.find(txn_id) != _pending_txn_ids.end();
+}
+
+StatusOr<int64_t> TxnStatePoller::TEST_pending_execution_time(int64_t txn_id) {
+    std::unique_lock<bthread::Mutex> lock(_mutex);
+    auto it = _pending_tasks.begin();
+    while (it != _pending_tasks.end()) {
+        if (it->second.txn_id == txn_id) {
+            return it->first;
+        }
+        ++it;
+    }
+    return Status::NotFound("no task found");
+}
+
+bool TxnStatePoller::TEST_is_scheduling() {
+    std::unique_lock<bthread::Mutex> lock(_mutex);
+    return _is_scheduling;
+}
+
+TxnStateCache::TxnStateCache(size_t capacity, std::unique_ptr<ThreadPoolToken> poller_token)
+        : _capacity(capacity), _poll_state_token(std::move(poller_token)) {
+    size_t capacity_per_shard = (_capacity + (kNumShards - 1)) / kNumShards;
+    for (int32_t i = 0; i < kNumShards; i++) {
+        _shards[i] = std::make_unique<TxnStateDynamicCache>(capacity_per_shard);
+    }
+}
+
+Status TxnStateCache::init() {
+    _txn_state_poller = std::make_unique<TxnStatePoller>(this, _poll_state_token.get());
+    RETURN_IF_ERROR(_txn_state_poller->init());
+    _txn_state_clean_thread = std::make_unique<std::thread>([this] { _txn_state_clean_func(); });
```

Comment:
not sure if this can be put into puller's scheduling thread, to save the usage of threads. just a random thought. or just throw the task into bthread executor used for the poller?

Reply:
Good advice. Actually we need a scheduled executor here. The poller implements part of schedule logic but not general. I think it should be something like [kudu threadpool](https://github.com/apache/kudu/blob/master/src/kudu/util/threadpool.h#L285) or [folly ScheduledExecutor](https://github.com/facebook/folly/blob/main/folly/executors/ScheduledExecutor.h) so that many components can benefit from it, such as many schedule threads in storage_engine.h. But I have no time to do it in this PR. let's leave it alone now.